mobile_verifier/src/subnetwork_rewards.rs

```diff
@@ -0,0 +1,687 @@
+use crate::{
+    bones_to_u64, cell_share_to_u64,
+    error::{Error, Result},
+    reward_share::{
+        self, cell_shares, hotspot_shares, GatheredShares, OwnerEmissions, OwnerResolver,
+    },
+    reward_speed_share::SpeedShare,
+    write_message,
+};
+use chrono::{DateTime, Utc};
+use futures::stream::{self, StreamExt};
+use helium_proto::services::{follower, Channel};
+use poc_store::{FileInfo, FileStore, FileType};
+use serde::Serialize;
+
+mod proto {
+    pub use helium_proto::services::poc_mobile::*;
+    pub use helium_proto::{SubnetworkReward, SubnetworkRewards};
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct SubnetworkRewards {
+    pub after_ts: u64,
+    pub before_ts: u64,
+    pub file_list: Vec<FileInfo>,
+    pub gathered_shares: GatheredShares,
+    pub cell_shares: reward_share::CellShares,
+    pub hotspot_shares: reward_share::HotspotShares,
+    pub owner_shares: reward_share::OwnerShares,
+    pub missing_owner_shares: reward_share::MissingOwnerShares,
+    pub rewards: Vec<proto::SubnetworkReward>,
+}
+
+impl SubnetworkRewards {
+    pub fn is_empty(&self) -> bool {
+        self.rewards.is_empty()
+    }
+
+    pub async fn from_period(
+        file_store: &FileStore,
+        mut follower_service: follower::Client<Channel>,
+        after_utc: DateTime<Utc>,
+        before_utc: DateTime<Utc>,
+    ) -> Result<Self> {
+        if before_utc <= after_utc {
+            tracing::error!(
+                "cannot reward future period, before: {:?}, after: {:?}",
+                before_utc,
+                after_utc
+            );
+            return Err(Error::NotFound("cannot reward future".to_string()));
+        }
+        let after_ts = after_utc.timestamp() as u64;
+        let before_ts = before_utc.timestamp() as u64;
+
+        let mut file_list = file_store
+            .list_all(FileType::CellHeartbeat, after_utc, before_utc)
+            .await?;
+        file_list.extend(
+            file_store
+                .list_all(FileType::CellSpeedtest, after_utc, before_utc)
+                .await?,
+        );
+
+        metrics::histogram!("verifier_server_processed_files", file_list.len() as f64);
+
+        let mut stream = file_store.source(stream::iter(file_list.clone()).map(Ok).boxed());
+
+        let gathered_shares = GatheredShares::from_stream(&mut stream, after_ts, before_ts).await?;
+
+        let cell_shares = cell_shares(&gathered_shares.shares);
+
+        let hotspot_shares = hotspot_shares(
+            &gathered_shares.shares,
+            &gathered_shares.speed_shares_moving_avg,
+        );
+
+        let (owner_shares, missing_owner_shares) = follower_service
+            .owner_shares(hotspot_shares.clone())
```

Comment:
why do hotspot shares need to be cloned here instead of just passing a reference to them for the duration of this calculation?

Reply:
we need to write hotspot shares to s3 later

Reply:
still learning the borrow checker but even though this is an async function (can't figure out where the `owner_shares` method is implemented on that follower client) but we're immediately awaiting on it and the _usage_ is effectively synchronous so is this just by virtue of the call being async that the compiler can't be sure a reference to `hotspot_shares` passed to the `owner_shares` method will return before the main function ends and we need to write it out to s3?
or for that matter the value's not being mutated in either case so couldn't we safely pass an immutable reference to both `owner_shares` and the s3 writer?
not saying this needs to change so much as embiggening my understanding